src/luna.man
.\" This file is part of the Lightweight Universal Network Analyzer (LUNA)
.\"
.\" Copyright (c) 2013, 2014 Fiona Klute
.\"
.\" LUNA is free software: you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" LUNA is distributed in the hope that it will be useful, but WITHOUT
.\" ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public
.\" License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with LUNA. If not, see <http://www.gnu.org/licenses/>.
.TH LUNA 1 2014-06-15 "LUNA" "LUNA Manual"

.SH NAME
luna \- generate and record packets to measure network performance

.SH SYNOPSIS
.B luna
-s|-c SERVER [OPTIONS]...

.SH DESCRIPTION
.P
LUNA, short for "Lightweight Universal Network Analyzer", is a system
to analyze network behavior under different network loads. LUNA can
generate UDP packets following configured distributions (client mode),
or receive and log them (server mode). Optionally, if requested by the
client, the server will echo packets back to the client for round trip
time measurements. The LUNA source provides tools based on GNU Octave
for parsing logs and plot results, but you can also create your own
analysis tools.

.SH OPTIONS

.IP \fB\-s\fR
.PD 0
.TP
.B \-\-server
Run in server mode. \fB\-s\fR and \fB\-c\fR are mutually exclusive.

.IP "\fB\-c SERVER\fR"
.PD 0
.TP
.B \-\-client=SERVER
Run in client mode, send to the specified server. The server can be
given by hostname, IPv4 or IPv6 address. \fB\-c\fR and \fB\-s\fR are
mutually exclusive.

.IP "\fB\-p PORT\fR"
.PD 0
.TP
.B \-\-port=PORT
Use the specified port. In server mode, \fBluna\fR binds this UDP
port for listening. In client mode, this is used as the destination
port.

.IP \fB\-4\fR
.PD 0
.TP
.B \-\-ipv4
Run in IPv4 only mode

.IP \fB\-6\fR
.PD 0
.TP
.B \-\-ipv6
Run in IPv6 only mode

.IP \fB\-T\fR
.PD 0
.TP
.B \-\-tsv\-output
Write recorded data as tab separated output. This is required to use
the analysis tools provided by LUNA, and strongly recommended if you
want to write your own. May become default behavior in the future.

.IP "\fB\-t SECONDS\fR"
.PD 0
.TP
.B \-\-time=SECONDS
Send for the given number of seconds (client mode only)

.IP "\fB\-g GENERATOR\fR"
.PD 0
.TP
.B \-\-generator=GENERATOR
Use the named generator (client mode only). Default is "static". See
\fBGENERATORS\fR below.

.IP "\fB\-a ARGS\fR"
.PD 0
.TP
.B \-\-generator-args=ARGS
Pass the specified arguments to the configured generator (see option
.BR \-g ).
Generator arguments must be written in the format
name=value,... Acceptable names and values depend on the
generator. Packet intervals are generally given in microseconds.

.IP \fB\-e\fR
.PD 0
.TP
.B \-\-echo
Request echo packets for round trip time measurements (client mode only)

.IP "\fB\-o FILE\fR"
.PD 0
.TP
.B \-\-output=FILE
Write recorded results to FILE instead of standard out

.TP
.B \-\-start-time=UNIXTIME
Start the transmission at the given time (client mode only). The time
must be in Unix time (seconds since 1970-01-01). If this option is not
used, the transmission is started as soon as possible. This can be
used to start transmissions on different hosts at the same time,
assuming their clocks are in sync. The first packet will be sent at
the given time plus the first delay returned by the generator used.

.TP
.B \-\-clock=(realtime|monotonic)
Set the clock to use. Two clocks are available, realtime and
monotonic, corresponting to \fBCLOCK_REALTIME\fR and
\fBCLOCK_MONOTONIC\fR as described in
.BR clock_gettime (3).
If \fB--start-time\fR is used, LUNA always uses \fBCLOCK_REALTIME\fR,
otherwise \fBCLOCK_MONOTONIC\fR is the default but can be changed
using this option.

.SH EXIT STATUS
.P
.B 0
if execution completed normally, non-zero in case of an error. In the
latter case,
.BR luna
will output an error message for the user. Error exit codes are
defined in
.BR luna.h .

.SH ENVIRONMENT
.P
Generators using random generators from the GNU Scientific Library
(libgsl) may be influenced by the environment variables evaluated in
.BR gsl_rng_env_setup (),
see section "Random number environment variables" in the libgsl
handbook.

.SH GENERATORS
.P
A \fBgenerator\fR is a component of the LUNA client that provides
sizes and intervals for sending packets. You can choose between some
pre-defined generators or write your own. Each generator must provide
a creation function that returns a \fBgenerator_t\fR variable as
defined in \fBgenerator.h\fR. The same header also provides useful
helper functions. Generators usually accept arguments from the command
line, see option \fB-a\fR.

.P
Currently available generators are:

.TP
.B static
Generate packets with a fixed size and interval, as defined by
generator arguments \fBsize\fR and
\fBinterval\fR.

.TP
.B random_size
Use packet sizes evenly distributed between the provided \fBsize\fR
argument and the minimum size required by the LUNA protocol. The
\fBinterval\fR argument is used as a static value.

.TP
.B alt_time
Alternate packet intervals between the provided \fBinterval\fR
argument and double that value every ten packets. The \fBsize\fR
argument is used as a static value. This generator is primarily an
example.

.TP
.B gaussian
This generator creates packets with sizes following a Gaussian
distribution, based on a random generator provided by
\fBlibgsl\fR. Three parameters can be configured: \fBmax\fR sets the
maximum permitted packet size, \fBsigma\fR is the standard deviation
for the Gaussian distribution, and \fBinterval\fR the static packet
interval. The distributions mean value will be half the
maximum. Random values above and below the permitted range will be
replaced with the maximum and minimum values, respectively. Behavior
and available arguments of this generator will likely change in the
future for more flexibility.

.P
Unless mentioned otherwise, all generators listed above use a default
packet interval of 1000µs.

.SH NOTES

.P
LUNA tries to acquire realtime priority and lock its memory into RAM
to optimize timing performance. This requires the \fBCAP_IPC_LOCK\fR
and \fBCAP_SYS_NICE\fR capabilities. The \fB./configure\fR script
provides the option \fB\-\-enable\-capabilities\fR to set file
capabilities at install time using
.BR setcap (8),
but whether using this option is appropriate depends on the system's
security policy, see
.BR capabilities (7)
for details. LUNA will work without these capabilities, but likely
with reduced performance.

.P
For optimal timing precision, it is recommended to use LUNA with a
\%PREEMPT_RT patched Linux kernel to avoid delays caused by kernel
threads as far as possible. Whether or not this is relevant for your
application depends on your system properties and precision
requirements. Testing precision before taking measurements for
research is highly recommended in any case. Note that network hardware
can have a significant impact on packet timings in the microseconds
range.

.SH EXAMPLE
.P
Send one packet with 400 byte payload every 200µs to localhost for 20
seconds:
.RS
.P
luna -c localhost -g static -a size=400,interval=200 -t 20
.RE

.P
Send packets to 192.0.2.7 with sizes following a Gaussian distribution
around 200 bytes with a standard deviation of 30 bytes and a maximum
packet size of 400 bytes:
.RS
.P
luna -c 192.0.2.7 -g gaussian -a max=400,sigma=30
.RE

.SH SEE ALSO
.P
.BR luna-control (1)
for a high level control tool that can automate multiple connections.
Please see the LUNA source code for analysis tools.

.P
.BR capabilities (7)


src/luna.c
/*
 * This file is part of the Lightweight Universal Network Analyzer (LUNA)
 *
 * Copyright (c) 2013 Fiona Klute
 *
 * LUNA is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * LUNA is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public
 * License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with LUNA. If not, see <http://www.gnu.org/licenses/>.
 */
#include <config.h>

#include <arpa/inet.h>
#include <errno.h>
#include <getopt.h>
#include <netinet/in.h>
#include <netdb.h>
#include <sched.h>
#include <stdio.h>
#include <string.h>
#include <sys/mman.h>
#include <sys/socket.h>
#include <sys/time.h>
#include <sys/types.h>
#include <time.h>
#include <unistd.h>

#include "luna.h"
#include "server.h"
#include "client.h"

/* POSIX requires a minimum range of 32 for priorities. Using the
 * minimum plus 20 seems reasonable to aquire a high priority without
 * blocking any high priority processes that might be running. The
 * client needs a higher priority than the server because the server's
 * time measurements rely on the kernel and thus it has no reason to
 * block the client. A client offset of 60 ensures that it can preempt
 * medium priority kernel threads under PREEMPT_RT, which run at a
 * priority of 50. */
#define CLIENT_PRIO_OFFSET 60
#define SERVER_PRIO_OFFSET 20

/* Option codes for long options, starting at 260 to be safely above
 * charcodes for short options */
#define OPT_START_TIME 260
#define OPT_CLOCK 261

/* valid command line options for getopt */
#define CLI_OPTS "sc:p:46Tt:g:a:eo:"
struct option long_opts[] = {
	{"server",	no_argument,		NULL,	's'},
	{"client",	required_argument,	NULL,	'c'},
	{"port",	required_argument,	NULL,	'p'},
	{"ipv4",	no_argument,		NULL,	'4'},
	{"ipv6",	no_argument,		NULL,	'6'},
	{"tsv-output",	no_argument,		NULL,	'T'},
	{"time",	required_argument,	NULL,	't'},
	{"generator",	required_argument,	NULL,	'g'},
	{"generator-args", required_argument,	NULL,	'a'},
	{"echo",	no_argument,		NULL,	'e'},
	{"output",	required_argument,	NULL,	'o'},
	{"start-time",	required_argument,	NULL,	OPT_START_TIME},
	{"clock",	required_argument,	NULL,	OPT_CLOCK},
	{NULL,		0,			NULL,	0}
};

void chkalloc(void *ptr, char *file, int line)
{
	if (ptr == NULL)
	{
		fprintf(stderr,
			"Could not allocate required memory in %s, line %i!\n",
			file, line);
		exit(EXIT_MEMFAIL);
	}
}



int check_pfaults(struct rusage *pre, struct rusage *post)
{
	if (pre->ru_majflt != post->ru_majflt
	    || pre->ru_minflt != post->ru_minflt)
		return 1;
	else
		return 0;
}



int touch_page(void *mem, size_t size)
{
	/* get page size (once) */
	static long page_size = 0;
	if (page_size == 0)
		page_size = sysconf(_SC_PAGESIZE);

	/* touch each page in mem */
	char *c = mem;
	for (int i = 0; i < size; i += page_size)
		c[i] = 0;
}



void printtimeres()
{
	struct timespec timeres = {0, 0};
	if (clock_getres(CLOCK_MONOTONIC, &timeres) == -1)
		perror("Could not get clock resolution");
	else
		fprintf(stderr, "Kernel clock resolution: %ld.%09lds\n",
			timeres.tv_sec, timeres.tv_nsec);
}



int main(int argc, char *argv[])
{
	struct addrinfo addrhints;
	addrhints.ai_family = AF_UNSPEC;
	addrhints.ai_socktype = SOCK_DGRAM;
	addrhints.ai_protocol = IPPROTO_UDP;
	//	addrhints.ai_flags = AI_V4MAPPED | AI_ADDRCONFIG;
	addrhints.ai_flags = AI_V4MAPPED;
	addrhints.ai_addrlen = 0;
	addrhints.ai_addr = NULL;
	addrhints.ai_canonname = NULL;
	addrhints.ai_next = NULL;

	int server = 0;
	int client = 0;
	int flags = SERVER_SIGTERM_EXIT;
	int time = 1;
	struct timespec start_time = {0, 0};
	clockid_t clk_id = CLOCK_MONOTONIC;
	int echo = 0;
	/* port, host and clock will be allocated by strdup if needed,
	 * free'd below. */
	char *port = NULL;
	char *host = NULL;
	char *clock = NULL;
	/* the packet generator to use and its arguments */
	char *generator = NULL;
	char *gen_args = NULL;
	/* file path to write output data to, will be overwritten */
	char *datafile = NULL;

	for (int opt = getopt_long(argc, argv, CLI_OPTS, long_opts, NULL);
	     opt != -1;
	     opt = getopt_long(argc, argv, CLI_OPTS, long_opts, NULL))
	{
		switch (opt)
		{
		case 'p':
			port = strdup(optarg);
			CHKALLOC(port);
			break;
		case 's': // act as server
			if (client != 0)
			{
				fprintf(stderr, "Select client or server mode, "
					"never both!\n");
				exit(EXIT_INVALID);
			}
			server = 1;
			break;
		case 'c': // act as client
			if (server != 0)
			{
				fprintf(stderr, "Select client or server mode, "
					"never both!\n");
				exit(EXIT_INVALID);
			}
			host = strdup(optarg);
			CHKALLOC(host);
			client = 1;
			break;
		case '4': // IPv4 only
			addrhints.ai_family = AF_INET;
			break;
		case '6': // IPv6 only
			addrhints.ai_family = AF_INET6;
			flags |= SERVER_IPV6_ONLY;
			break;
		case 'T': // tab separated value output
			flags |= SERVER_TSV_OUTPUT;
			break;
		case 't': // time to run (s, client only)
			time = atoi(optarg);
			break;
		case 'g': // generator to use (client only)
			generator = strdup(optarg);
			CHKALLOC(generator);
			break;
		case 'a': // arguments for the generator (client only)
			gen_args = strdup(optarg);
			CHKALLOC(gen_args);
			break;
		case 'e': // request echo (client only)
			echo = 1;
			break;
		case 'o': // data output file
			datafile = strdup(optarg);
			CHKALLOC(datafile);
			break;
		case OPT_START_TIME:
			start_time.tv_sec = strtoul(optarg, NULL, 10);
			clk_id = CLOCK_REALTIME;
			break;
		case OPT_CLOCK:
			clock = strdup(optarg);
			CHKALLOC(clock);
			break;
		default:
			break;
		}
	}

	if (!(flags & SERVER_TSV_OUTPUT))
		printtimeres();

	if (port == NULL)
	{
		port = malloc(DEFAULT_PORT_LEN);
		CHKALLOC(port);
		snprintf(port, DEFAULT_PORT_LEN, "%u", DEFAULT_PORT);
	}

	if (generator == NULL)
	{
		generator = strdup(DEFAULT_GENERATOR);
		CHKALLOC(generator);
	}

	/* If true, the clock to use was set manually */
	if (clock != NULL)
	{
		if (strcmp(clock, "realtime") == 0)
			clk_id = CLOCK_REALTIME;
		else if (strcmp(clock, "monotonic") == 0)
			clk_id = CLOCK_MONOTONIC;
		else
		{
			fprintf(stderr, "Invalid clock: \"%s\"!\n", clock);
			exit(EXIT_INVALID);
		}
		free(clock);

		/* If one of the elements of start_time is non-zero,
		 * it was specified on the command line. Right now,
		 * only full seconds are possible, but by checking for
		 * nanoseconds, too, the code is prepared for higher
		 * precision. */
		if ((start_time.tv_sec != 0 || start_time.tv_nsec != 0)
		    && clk_id == CLOCK_MONOTONIC)
		{
			fprintf(stderr, "Error: Using a fixed start time is "
				"not possible with the monotonic clock!\n");
			exit(EXIT_INVALID);
		}
	}

	if (server)
	{
		addrhints.ai_flags |= AI_PASSIVE;
		if (addrhints.ai_family != AF_INET)
			addrhints.ai_family = AF_INET6;
	}
	else
		if (host == NULL)
		{
			fprintf(stderr, "You must either use server mode or "
				"specify a server to send to (-c HOST)!\n");
			exit(EXIT_INVALID);
		}

	/* res will be allocated by getaddrinfo and free'd in
	 * client/server functions. */
	struct addrinfo *res = NULL;
	int error = 0;
	error = getaddrinfo(host, port, &addrhints, &res);
	if (error != 0)
	{
		fprintf(stderr, "Error in getaddrinfo for \"%s\": %s\n",
			host, gai_strerror(error));
	}

	free(host);
	free(port);

	int retval = 0;
	/* lock all allocated memory into RAM before starting realtime
	 * sections (needs capability CAP_IPC_LOCK) */
	if (mlockall(MCL_CURRENT | MCL_FUTURE))
	{
		perror("mlockall failed");
		fprintf(stderr, "Make sure LUNA has the CAP_IPC_LOCK "
			"capability! If you don't need real-time "
			"precision, you can safely ignore this warning.\n");
	}

	/* Try to get real-time priority */
	struct sched_param sparam;
	memset(&sparam, 0, sizeof(struct sched_param));
	if (client)
		sparam.sched_priority =
			sched_get_priority_min(SCHED_RR) + CLIENT_PRIO_OFFSET;
	else
		sparam.sched_priority =
			sched_get_priority_min(SCHED_RR) + SERVER_PRIO_OFFSET;
	if (sched_setscheduler(0, SCHED_RR, &sparam) == -1)
	{
		perror("Could not get real-time priority");
		fprintf(stderr, "Make sure LUNA has the CAP_SYS_NICE "
			"capability! If you don't need real-time "
			"precision, you can safely ignore this warning.\n");
	}

	if (client)
		retval = run_client(res, time,
				    start_time, clk_id,
				    echo,
				    generator, gen_args,
				    datafile);
	free(gen_args);
	free(generator);

	if (server)
		retval = run_server(res, flags, datafile);
	free(datafile);

	return retval;
}
